ambari-agent/src/main/python/ambari_agent/FileCache.py

```diff
@@ -74,12 +74,15 @@ def get_service_base_dir(self, command, server_url_prefix):
     """
     Returns a base directory for service
     """
-    if 'service_package_folder' in command['commandParams']:
-      service_subpath = command['commandParams']['service_package_folder']
-    else:
-      service_subpath = command['serviceLevelParams']['service_package_folder']
-    return self.provide_directory(self.cache_dir, service_subpath,
-                                  server_url_prefix)
+    key = 'service_package_folder'
+    service_subpath = None
+    for param_type in ('commandParams', 'serviceLevelParams'):
+      if param_type in command and key in command[param_type]:
+        service_subpath = command[param_type][key]
+        break
+
+    if service_subpath:
```

Comment:
I think "if else it much' readable than your approach. Considering it's only two dictionaries.
 
Also KeyError looks a better indicator of what went wrong than NPE (in your case). Meaning function will return None and something later will fail with NPE - very hard to debug

Reply:
1. There is much less duplication using this approach than with if-else, and less prone to typos.  I think it is more readable, too, due to reduced clutter.
2. In general I agree, but `get_hook_base_dir` and `get_custom_resources_subdir` also `return None`.  Are those "optional"?

Reply:
I am with @aonishuk on the readability issue. It took me a few reads to figure out what was going on.  This is mainly due to the fact that I am not familiar with the code - as is probably the case for more developers.

Reply:
Either this implementation or another with 'if-else' conditions should go out ASAP since it blocks other PRs to be merged.